File: Daily Resources/Day01/DSA - Binary Search Solutions/Array Division.py
def is_valid(nums, k, max_sum):
    count = 1
    current = 0
    for num in nums:
        if current + num > max_sum:
            count += 1
            current = num
            if count > k:
                return False
        else:
            current += num
    return True

def find_min_largest_sum(nums, k):
    low = max(nums)
    high = sum(nums)
    result = high

    while low <= high:
        mid = (low + high) // 2
        if is_valid(nums, k, mid):
            result = mid
            high = mid - 1
        else:
            low = mid + 1
    return result

# Read input
n, k = map(int, input().split())
nums = list(map(int, input().split()))

# Output result
print(find_min_largest_sum(nums, k))

File: Daily Resources/Day01/DSA - Binary Search Solutions/Array Division.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

bool isValid(const vector<long long>& arr, int k, long long maxSum) {
    long long currentSum = 0;
    int count = 1;

    for (auto val : arr) {
        if (currentSum + val > maxSum) {
            currentSum = val;
            count++;
            if (count > k) return false;
        } else {
            currentSum += val;
        }
    }
    return true;
}

int main() {
    int n, k;
    cin >> n >> k;
    vector<long long> arr(n);
    long long maxVal = 0, totalSum = 0;

    for (int i = 0; i < n; ++i) {
        cin >> arr[i];
        maxVal = max(maxVal, arr[i]);
        totalSum += arr[i];
    }

    long long left = maxVal, right = totalSum, result = totalSum;

    while (left <= right) {
        long long mid = (left + right) / 2;
        if (isValid(arr, k, mid)) {
            result = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    cout << result << endl;
    return 0;
}
